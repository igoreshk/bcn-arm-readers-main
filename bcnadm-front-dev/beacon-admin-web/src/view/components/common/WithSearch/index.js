import React, { Component } from 'react';
import PropTypes from 'prop-types';

const defaultContext = {
  filter: '',
  setFilter: () => {
    throw new Error('SearchPanel component should only be used in a component wrapped with WithSearch');
  }
};

export const WithSearchContext = React.createContext(defaultContext);

const WithSearch = (WrappedComponent) =>
  class extends Component {
    static displayName = `WithSearch(${WrappedComponent.displayName || WrappedComponent.name || 'Component'})`;
    static propTypes = {
      keys: PropTypes.arrayOf(PropTypes.string).isRequired,
      source: PropTypes.arrayOf(PropTypes.object).isRequired
    };
    constructor(props) {
      super(props);
      this.state = {
        data: this.props.source,
        filter: ''
      };
    }
    componentDidUpdate(prevProps, prevState) {
      if (prevState.data === this.state.data) {
        this.applyFilter();
      }
    }

    applyFilter() {
      const {
        state: { filter },
        props: { keys, source }
      } = this;
      this.setState({
        data: source.filter(
          (object) => keys.filter((key) => object[key].toLowerCase().includes(filter.toLowerCase())).length > 0
        )
      });
    }

    setFilter = ({ target: { value: filter } }) => {
      this.setState({ filter }, this.applyFilter);
    };

    render() {
      return (
        <WithSearchContext.Provider value={{ filter: this.state.filter, setFilter: this.setFilter }}>
          <WrappedComponent {...this.state} {...this.props} />
        </WithSearchContext.Provider>
      );
    }
  };

export default WithSearch;
