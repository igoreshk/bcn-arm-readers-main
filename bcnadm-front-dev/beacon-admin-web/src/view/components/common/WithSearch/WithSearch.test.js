import React from 'react';
import { shallow } from 'enzyme';
import WithSearch, { WithSearchContext } from './index';

const props = {
  keys: ['key_1', 'key_2', 'key_3'],
  source: [
    {
      key_1: 'key_1',
      key_2: 'key_2',
      key_3: 'key_3'
    },
    {
      key_1: 'key_2',
      key_2: 'key_3',
      key_3: 'key_4'
    }
  ]
};

class ConsumerComponent extends React.Component {
  render() {
    return (
      <WithSearchContext.Consumer>
        {({ filter, setFilter }) => (
          <div>
            <div>Hello World</div>
            <input type="text" onChange={setFilter} value={filter} />
          </div>
        )}
      </WithSearchContext.Consumer>
    );
  }
}
const WrappedComponent = WithSearch(ConsumerComponent);
const wrapper = shallow(<WrappedComponent {...props} />);
const applyFilter = jest.spyOn(WrappedComponent.prototype, 'applyFilter');
const mockEvent = { target: { value: 'key_1' } };

describe('WithSearch', () => {
  it('Should check that props are passed properly', () => expect(wrapper.instance().props).toEqual(props));

  it('Should apply filter value to data array', () => {
    wrapper.instance().setFilter(mockEvent);
    expect(wrapper.instance().state.data[0]).toEqual(props.source[0]);
    expect(wrapper.instance().state.data).toHaveLength(1);
    expect(applyFilter).toHaveBeenCalled();
  });

  it('Should match snapshot', () => expect(wrapper).toMatchSnapshot());
});
