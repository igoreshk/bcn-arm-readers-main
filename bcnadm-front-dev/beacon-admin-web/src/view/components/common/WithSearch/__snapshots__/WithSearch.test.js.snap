// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithSearch Should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "filter": "key_1",
      "setFilter": [Function],
    }
  }
>
  <ConsumerComponent
    data={
      Array [
        Object {
          "key_1": "key_1",
          "key_2": "key_2",
          "key_3": "key_3",
        },
      ]
    }
    filter="key_1"
    keys={
      Array [
        "key_1",
        "key_2",
        "key_3",
      ]
    }
    source={
      Array [
        Object {
          "key_1": "key_1",
          "key_2": "key_2",
          "key_3": "key_3",
        },
        Object {
          "key_1": "key_2",
          "key_2": "key_3",
          "key_3": "key_4",
        },
      ]
    }
  />
</ContextProvider>
`;
